### ISSUE TITLE
Eliminate arbitrary token types (not Type, not OpaqueToken)

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/angular#901]
From an internal customer after https://github.com/dart-lang/angular/pull/898/commits/6d37665ac2ba166727c6720994a887321797227d, they hit:

> ```
> Error: Constructor is not a 'const' constructor.
>     const [const _ngRef.Inject(const SomeToken())],
>                                      ^^^^^^^^^^
> ```

This is a bug. We shouldn't have impacted existing code. They had code like this:

```dart
/// Annotation used to qualify [SomeToken] value.
class SomeToken {
  const SomeToken();
}

var providers = [
   new Provider(prefix.SomeToken, useValue: someValue),
];

@Injectable()
class Service {
  Service(@Inject(prefix.SomeToken) var someValue) { ... }
}
```

**Comment by matanlurey** [comment]
Turns out this is just an invalid use of this feature. We should break earlier, though.

**Comment by matanlurey** [comment]
Internal CL: cr/186016325.

**Comment by matanlurey** [comment]
Just an update: looking pretty good.

I'm adding =5.0.0 because it looks safe to get in by 5.0.0-beta.

**Comment by matanlurey** [comment]
Down to the last 1-2 uses internally.

I'll start on the PR that removes the test cases and throws instead.

**Comment by matanlurey** [comment]
This is fixed now as of #1054!
